File: mobile-app/lib/data/weather/open_meteo_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../domain/entities/companion_models.dart';
import 'weather_cities.dart';

class OpenMeteoClient {
  OpenMeteoClient({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;
  DateTime? _lastFetchAt;
  WeatherReading? _cached;

  Future<WeatherReading?> fetch(WeatherCity city) async {
    final now = DateTime.now().toUtc();
    if (_cached != null &&
        _lastFetchAt != null &&
        now.difference(_lastFetchAt!) < const Duration(minutes: 15)) {
      return _cached;
    }

    final uri =
        Uri.https('api.open-meteo.com', '/v1/forecast', <String, String>{
          'latitude': city.latitude.toString(),
          'longitude': city.longitude.toString(),
          'current': 'temperature_2m,precipitation',
          'hourly': 'precipitation_probability',
          'forecast_hours': '1',
          'timezone': 'GMT',
        });

    try {
      final response = await _client
          .get(uri)
          .timeout(const Duration(seconds: 8));
      if (response.statusCode != 200) {
        return _staleOrNull();
      }
      final json = jsonDecode(response.body) as Map<String, Object?>;
      final current = json['current'] as Map<String, Object?>?;
      if (current == null) return _staleOrNull();

      final temp = (current['temperature_2m'] as num?)?.toDouble();
      final precipitation = (current['precipitation'] as num?)?.toDouble() ?? 0;
      final hourly = json['hourly'] as Map<String, Object?>?;
      final popList = hourly?['precipitation_probability'] as List<Object?>?;
      final pop = popList == null || popList.isEmpty
          ? CompanionSnapshot.popInvalid
          : (popList.first as num).round().clamp(0, 100);

      if (temp == null) return _staleOrNull();

      final reading = WeatherReading(
        tempCX10: (temp * 10).round(),
        popPct: pop,
        rainNow: precipitation > 0,
        rainSoon: pop >= 30,
        stale: false,
      );
      _cached = reading;
      _lastFetchAt = now;
      return reading;
    } on Object {
      return _staleOrNull();
    }
  }

  WeatherReading? _staleOrNull() {
    if (_cached == null) return null;
    return WeatherReading(
      tempCX10: _cached!.tempCX10,
      popPct: _cached!.popPct,
      rainNow: _cached!.rainNow,
      rainSoon: _cached!.rainSoon,
      stale: true,
    );
  }
}

File: web-app/src/services/openMeteoClient.ts
import { CompanionSnapshotConsts, type WeatherReading } from '../domain/types';
import type { WeatherCity } from './weatherCities';

export class OpenMeteoClient {
  private lastFetchAt: Date | null = null;
  private cached: WeatherReading | null = null;

  async fetch(city: WeatherCity): Promise<WeatherReading | null> {
    const now = new Date();
    if (
      this.cached &&
      this.lastFetchAt &&
      now.getTime() - this.lastFetchAt.getTime() < 15 * 60 * 1000
    ) {
      return this.cached;
    }

    const params = new URLSearchParams({
      latitude: String(city.latitude),
      longitude: String(city.longitude),
      current: 'temperature_2m,precipitation',
      hourly: 'precipitation_probability',
      forecast_hours: '1',
      timezone: 'GMT',
    });
    try {
      const response = await fetch(`https://api.open-meteo.com/v1/forecast?${params}`, {
        signal: AbortSignal.timeout(8000),
      });
      if (!response.ok) return this.staleOrNull();
      const json = (await response.json()) as Record<string, unknown>;
      const current = json.current as Record<string, unknown> | undefined;
      if (!current) return this.staleOrNull();
      const temp = current.temperature_2m as number | undefined;
      const precipitation = (current.precipitation as number | undefined) ?? 0;
      const hourly = json.hourly as Record<string, unknown> | undefined;
      const popList = hourly?.precipitation_probability as number[] | undefined;
      const pop =
        !popList || popList.length === 0
          ? CompanionSnapshotConsts.popInvalid
          : Math.max(0, Math.min(100, Math.round(popList[0])));

      if (temp === undefined) return this.staleOrNull();
      const reading: WeatherReading = {
        tempCX10: Math.round(temp * 10),
        popPct: pop,
        rainNow: precipitation > 0,
        rainSoon: pop >= 30,
        stale: false,
      };
      this.cached = reading;
      this.lastFetchAt = now;
      return reading;
    } catch {
      return this.staleOrNull();
    }
  }

  private staleOrNull(): WeatherReading | null {
    if (!this.cached) return null;
    return { ...this.cached, stale: true };
  }
}
